src/lib.rs

```diff
@@ -71,98 +71,14 @@ mod jwt;
 mod util;
 
-use custom_service_account::CustomServiceAccount;
-
 pub use authentication_manager::AuthenticationManager;
+pub use custom_service_account::CustomServiceAccount;
 pub use error::Error;
-pub use types::Token;
-
-use std::path::Path;
-
-use hyper::Client;
-use hyper_rustls::HttpsConnectorBuilder;
-
-/// Initialize GCP authentication based on a credentials file path
-///
-/// Returns `AuthenticationManager` which can be used to obtain tokens
-pub async fn from_credentials_file<T: AsRef<Path>>(
-    path: T,
-) -> Result<AuthenticationManager, Error> {
-    let custom = CustomServiceAccount::from_file(path.as_ref()).await?;
-    get_authentication_manager(Some(custom)).await
-}
+pub use types::{Signer, Token};
 
-/// Initialize GCP authentication based on a JSON string
-///
-/// Returns `AuthenticationManager` which can be used to obtain tokens
-pub async fn from_credentials_json(s: &str) -> Result<AuthenticationManager, Error> {
-    let custom = CustomServiceAccount::from_json(s)?;
-    get_authentication_manager(Some(custom)).await
-}
-
-async fn get_authentication_manager(
-    custom: Option<CustomServiceAccount>,
-) -> Result<AuthenticationManager, Error> {
-    #[cfg(feature = "webpki-roots")]
-    let https = HttpsConnectorBuilder::new().with_webpki_roots();
-    #[cfg(not(feature = "webpki-roots"))]
-    let https = HttpsConnectorBuilder::new().with_native_roots();
-
-    let client =
-        Client::builder().build::<_, hyper::Body>(https.https_or_http().enable_http2().build());
-
-    if let Some(service_account) = custom {
-        log::debug!("Using CustomServiceAccount");
-        return Ok(AuthenticationManager::new(
-            client,
-            Box::new(service_account),
-        ));
-    }
-    let gcloud = gcloud_authorized_user::GCloudAuthorizedUser::new().await;
-    if let Ok(service_account) = gcloud {
-        log::debug!("Using GCloudAuthorizedUser");
-        return Ok(AuthenticationManager::new(
-            client.clone(),
-            Box::new(service_account),
-        ));
-    }
-    let default = default_service_account::DefaultServiceAccount::new(&client).await;
-    if let Ok(service_account) = default {
-        log::debug!("Using DefaultServiceAccount");
-        return Ok(AuthenticationManager::new(
-            client.clone(),
-            Box::new(service_account),
-        ));
-    }
-    let user = default_authorized_user::DefaultAuthorizedUser::new(&client).await;
-    if let Ok(user_account) = user {
-        log::debug!("Using DefaultAuthorizedUser");
-        return Ok(AuthenticationManager::new(client, Box::new(user_account)));
-    }
-    Err(Error::NoAuthMethod(
-        Box::new(gcloud.unwrap_err()),
-        Box::new(default.unwrap_err()),
-        Box::new(user.unwrap_err()),
-    ))
-}
 /// Initialize GCP authentication
 ///
-/// Returns `AuthenticationManager` which can be used to obtain tokens
+/// This is a convenience wrapper around [`AuthenticationManager::new()`].
```

Comment:
I personally like the shortcut of being able to do `gcp_auth::init()` without doing any imports, especially in helper CLIs. Seems like a cheap way to get a slicely nicer UX.

Reply:
I have no strong opinions and recognise this is entirely bikeshedding, so if you like to keep both that's perfectly fine.  Though in rust I did get very used to the `SomeStruct::constructor()` way over an init shortcut and when I have the option (I think tracing_subscriber::fmt() is another example) I will still choose the constructor with the struct name in it as it is more descriptive to me to see the struct name being created.

Reply:
Interesting data point! I think I rather like the import-less initialization aspect to me but I will recognize it's less idiomatic then calling a type constructor.

Again, since @hrvolapeter came up with this API initially I'll let them weigh in before I make any further changes.

Reply:
the whole init think is probably just a habit from university, we've been taught that constructor shouldn't have side effects due to weird error handling in C++ constructors which is not really a valid point in rust anymore. `init` for me is basically constructor with side effects, I think similar approach is quite common in OOP but agree that it probably doesn't look the most rust native way.